File: labltk/src/dance.ml
(*
https://web.archive.org/web/20140122124421/http:/ensign.editme.com/t43dances

Usage:
opam install -y labltk

 *)

type planet_t =
  | Mercury
  | Venus
  | Earth
  | Mars
  | Jupiter
  | Saturn
  | Uranus
  | Neptune
  | Pluto
;;

let year = function
  | Mercury -> 87.969
  | Venus -> 224.701
  | Earth -> 365.256
  | Mars -> 686.980
  | Jupiter -> 4332.6
  | Saturn -> 10759.2
  | Uranus -> 30685.
  | Neptune -> 60190.
  | Pluto -> 90465.
;;

let orbit = function
  | Mercury -> 57.91
  | Venus -> 108.21
  | Earth -> 149.60
  | Mars -> 227.92
  | Jupiter -> 778.57
  | Saturn -> 1433.5
  | Uranus -> 2872.46
  | Neptune -> 4495.1
  | Pluto -> 5869.7
;;

let name = function
  | Mercury -> "Mercury"
  | Venus -> "Venus"
  | Earth -> "Earth"
  | Mars -> "Mars"
  | Jupiter -> "Jupiter"
  | Saturn -> "Saturn"
  | Uranus -> "Uranus"
  | Neptune -> "Neptune"
  | Pluto -> "Pluto"
;;

(* let draw_caption outer_planet inner_planet orbits =
 *   let outer_planet_name = name outer_planet in
 *   let inner_planet_name = name inner_planet in
 *   let orbit_text = Printf.sprintf "%.0f" orbits in
 *   draw_text 10 10 *)
let canvas_len = 400
let pi = 4.0 *. atan 1.0

let black = `Color "#000000"
let blue = `Color "#0000FF"
let red = `Color "#FF0000"
let green = `Color "#00FF00"
let purple = `Color "#800080"
let maroon = `Color "#800000"
let navy = `Color "#8B0000"
let dark_red = `Color "#8B0000"
let orange = `Color "#FFA500"

let draw_text canvas color x y text =
  Canvas.create_text ~fill:color ~text:text ~x ~y canvas
;;
           
let draw_line canvas color from_x from_y to_x to_y =
  Canvas.create_line ~fill:color ~xys:[(from_x, from_y); (to_x, to_y)] canvas
;;

let main canvas outer_planet inner_planet orbits =
  let outer_planet_year = year outer_planet in
  let inner_planet_year = year inner_planet in
  let outer_planet_radius = orbit outer_planet in
  let inner_planet_radius = orbit inner_planet in
  let interval_days = outer_planet_year /. 75. in
  let ycenter = float_of_int (canvas_len / 2) in
  let xcenter = float_of_int (canvas_len / 2) in
  (* outer radius *)
  let r1 = ycenter in 
  let r2 = r1 *. inner_planet_radius /. outer_planet_radius in
  let r = ref 0. in
  let rstop = outer_planet_year *. orbits in
  let a1 = ref 0. in
  let a1_interval = 2. *. pi *. interval_days /. outer_planet_year in
  let a2 = ref 0. in
  let a2_interval = 2. *. pi *. interval_days /. inner_planet_year in
  (* wait until q key is pressed*)
  begin
    (* draw text *)
    let outer_planet_name = name outer_planet in
    let inner_planet_name = name inner_planet in
    let orbit_text = Printf.sprintf "%.0f orbits" orbits in
    ignore @@ draw_text canvas blue 30 20 outer_planet_name;
    ignore @@ draw_text canvas blue 30 40 inner_planet_name;
    ignore @@ draw_text canvas blue 30 (canvas_len - 20) orbit_text;
    while !r < rstop do
      let i = int_of_float (floor (!r /. interval_days /. 75.)) in
      let color = 
        if i = 0 then black
        else if i = 1 then blue
        else if i = 2 then red
        else if i = 3 then green
        else if i = 4 then purple
        else if i = 5 then maroon
        else if i = 6 then navy
        else if i = 7 then dark_red
        else orange in
      begin
        a1 := !a1 -. a1_interval;
        a2 := !a2 -. a2_interval;
        let x1 = r1 *. cos !a1 in
        let y1 = r1 *. sin !a1 in
        let x2 = r2 *. cos !a2 in
        let y2 = r2 *. sin !a2 in
        ignore @@ draw_line canvas color
          (int_of_float (x1 +. xcenter)) (int_of_float (y1 +. ycenter))
          (int_of_float (x2 +. xcenter)) (int_of_float (y2 +. ycenter));
        r := !r +. interval_days;
      end;
    done;
  end
;;

let _ =
  let window = Tk.openTk () in
  let canvas = Canvas.create ~width:canvas_len ~height:canvas_len window in
  Tk.pack [canvas];
  main canvas Earth Venus 8.;
  Printexc.print Tk.mainLoop ()
;;

File: java/src/main/java/com/mamewo/Dance.java
package com.mamewo;

import java.awt.BasicStroke;
import java.awt.Canvas;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.Point;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.awt.event.MouseMotionListener;
import java.awt.image.BufferedImage;

import javax.swing.JFrame;
import javax.swing.JPanel;

public class Dance {
    static private int w = 400, h = 400;
    static private Color BLACK = new Color(0x00, 0x00, 0x00);
    static private Color BLUE = new Color(0x00,0x00,0xFF);
    static private Color RED = new Color(0xFF,0x00,0x00);
    static private Color GREEN = new Color(0x00,0xFF,0x00);
    static private Color PURPLE = new Color(0x80,0x00,0x80);
    static private Color MAROON = new Color(0x80,0x00,0x00);
    static private Color NAVY = new Color(0x8B,0x00,0x00);
    static private Color DARK_RED = new Color(0x8B,0x00,0x00);
    static private Color ORANGE = new Color(0xFF,0xA5,0x00);
        
    private enum Planet {
        Mercury,
        Venus,
        Earth,
        Mars,
        Jupiter,
        Saturn,
        Uranus,
        Neptune,
        Pluto
    };
    
    private static double year(Planet planet){
        switch(planet){
        case Mercury:
            return 87.969;
        case Venus:
            return 224.701;
        case Earth:
            return 365.256;
        case Mars:
            return 686.980;
        case Jupiter:
            return 4332.6;
        case Saturn:
            return 10759.2;
        case Uranus:
            return 30685.0;
        case Neptune:
            return 60190.0;
        case Pluto:
            return 90465.0;
        default:
            assert false;
        }
        return -1.0;
    }

    private static double orbit(Planet planet){
        switch(planet){
        case Mercury: return 57.91;
        case Venus: return 108.21;
        case Earth: return 149.60;
        case Mars: return 227.92;
        case Jupiter: return 778.57;
        case Saturn: return 1433.5;
        case Uranus: return 2872.46;
        case Neptune: return 4495.1;
        case Pluto: return 5869.7;
        default:
            assert false;
        }
        return -1.0;
    }

    public static void main(String[] args) {
        run();
    }

    public static void run() {
        final double orbits = 8.0;
        final int headerHeight = 30;
        // JFrameのインスタンスを生成
        JFrame frame = new JFrame("Dance of the planets");
        // ウィンドウを閉じたらプログラムを終了する
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        // ウィンドウのサイズ・初期位置
        frame.setSize(w, h + headerHeight);
        frame.setLocationRelativeTo(null);

        // PaintCanvasのインスタンスを生成
        PaintCanvas canvas = new PaintCanvas(Planet.Earth, Planet.Venus, orbits);

        // フレームに追加
        JPanel pane = new JPanel();
        frame.getContentPane().add(pane);

        canvas.setPreferredSize(new Dimension(w, h));
        pane.add(canvas);

        // ウィンドウを表示
        frame.setVisible(true);
    }

    // キャンバスクラス
    static class PaintCanvas extends Canvas
    {
        private final Planet outerPlanet_;
        private final Planet innerPlanet_;
        private final double orbits_;

        public PaintCanvas(Planet outerPlanet, Planet innerPlanet, double orbits) {
            // 座標を初期化
            outerPlanet_ = outerPlanet;
            innerPlanet_ = innerPlanet;
            orbits_ = orbits;

            // キャンバスの背景を白に設定
            setBackground(Color.white);

            // 描画
            repaint();
        }

        @Override
        public void paint(Graphics g) {
            Graphics2D g2d = (Graphics2D)g;
            final double outerPlanetYear = year(outerPlanet_);
            final double innerPlanetYear = year(innerPlanet_);
            final double outerPlanetRadius = orbit(outerPlanet_);
            final double innerPlanetRadius = orbit(innerPlanet_);
            final double intervalDays = outerPlanetYear / 75.0;
            final int ycenter = w / 2;
            final int xcenter = h / 2;
            final double r1 = ycenter;
            final double r2 = r1 * innerPlanetRadius / outerPlanetRadius;
            double r = 0.0;
            final double rstop = outerPlanetYear * orbits_;
            double a1 = 0.0;
            double a1Interval = 2.0 * Math.PI * intervalDays / outerPlanetYear;
            double a2 = 0.0;
            final double a2Interval = 2 * Math.PI * intervalDays / innerPlanetYear;
            final String outerPlanetName = outerPlanet_.toString();
            final String innerPlanetName = innerPlanet_.toString();

            while(r < rstop){
                int i = (int)(Math.floor (r / intervalDays / 75.0));
                Color color;
                if(i == 0){
                    color = BLACK;
                } else if(i == 1){
                    color = BLUE;
                } else if(i == 2){
                    color = RED;
                } else if(i == 3){
                    color = GREEN;
                } else if(i == 4){
                    color = PURPLE;
                } else if(i == 5){
                    color = MAROON;
                } else if(i == 6){
                    color = NAVY;
                } else if(i == 7){
                    color = DARK_RED;
                } else {
                    color = ORANGE;
                }
                a1 = a1 - a1Interval;
                a2 = a2 - a2Interval;
                final int x1 = (int)(r1 * Math.cos(a1));
                final int y1 = (int)(r1 * Math.sin(a1));
                final int x2 = (int)(r2 * Math.cos(a2));
                final int y2 = (int)(r2 * Math.sin(a2));
                g2d.setColor(color);
                g2d.drawLine(x1 + xcenter, y1 + ycenter, x2  + xcenter, y2 + ycenter);
                r = r + intervalDays;
            }
        }
    }
}
